### PR TITLE
Implement single_branch flag on git_repository rule

### PR DESCRIPTION
**Pull request by kerrick-lyft**
This allows passing the `--single-branch` flag to the `git clone` command, which can make clones faster if there are many branches in the repo.

### CODE REVIEW

**Comment by jin** [comment]
Adding @aehlig since @dslomov is unavailable temporarily

**Comment by edbaunton** [comment]
As a curiosity, is there a limitation that prevents one from specifying `--single-branch` in all cases where a branch is specified?

Specifically, [here](https://github.com/bazelbuild/bazel/blob/91493dc017bfbd39313e60ced95228b6c86ecef1/tools/build_defs/repo/git.bzl#L31-L32) we could tack on the `--single-branch` to speed up clones where a branch is specifically set.

File: tools/build_defs/repo/git.bzl [hunk 1]
```diff
@@ -55,7 +58,7 @@ set -ex
 ( cd {working_dir} &&
     if ! ( cd '{dir_link}' && [[ "$(git rev-parse --git-dir)" == '.git' ]] ) >/dev/null 2>&1; then
       rm -rf '{directory}' '{dir_link}'
-      git clone '{shallow}' '{remote}' '{directory}' || git clone '{remote}' '{directory}'
+      git clone '{shallow}' '{single_branch}' '{remote}' '{directory}' || git clone '{remote}' '{directory}'
```
Comment (aehlig) [inline tools/build_defs/repo/git.bzl]: Here I'm confused about the quoting. Your `single_branch` variable brings to options (`-b` and `--single-branch`) and should interpolate as 3 `argv` entries to `git`. However, with the given single quotes, those form a single argument.

**Comment by aehlig** [comment]
> Specifically, [here](https://github.com/bazelbuild/bazel/blob/91493dc017bfbd39313e60ced95228b6c86ecef1/tools/build_defs/repo/git.bzl#L31-L32) we could tack on the `--single-branch` to speed up clones where a branch is specifically set.

There we already set `--depth=1` which imples `--single-branch` unless an explicit `--no-single-branch` is given (which is not).

**Comment by kerrick-lyft** [comment]
@aehlig sorry for the late reply; I was on vacation.

Good catch on the quoting issue, I fixed it. I checked our internal fork of `git.bzl` and we did not have the quotes in that copy; I forget why I added them to this PR.

Given that this issue made it into this PR I will add some unit tests to make sure that this PR is correct. I will do that later today and @-mention you on the PR once it's done.

**Comment by kerrick-lyft** [comment]
Update: I believe this is obsoleted by the changes in https://github.com/bazelbuild/bazel/commit/9c913b93e275038393e4ba5f67b7421919d93cf9. With that commit, `git_repository` now does a fetch rather than a clone, which means it avoids pulling all branches (achieving the same thing as this PR wants to do).

I will close this PR for now. Once we've internally updated to a Bazel version that includes that commit (and if I remember) I will update this PR on whether the change in 9c913b9 solves our problem.

**Pull request closed without merging**